### PR TITLE
Endorsement/Reputation/Bucketing

### PR DESCRIPTION
**Pull request by carlaKC**

### CODE REVIEW

File: recommendations/local-resource-conservation.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+* [Implementation Notes](#implementation-notes)
+  * [Decaying Average](#decaying-average)
+  * [Multiple Channels](#multiple-channels)
+
+## Introduction
+Channel jamming is a known denial-of-service attack against the Lightning 
+Network. An attacker that controls both ends of a payment route can disrupt 
+usage of a channel by sending payments that are destined to fail through a 
+target route. This attack can exhaust the resources along the target route in 
+two ways, though the difference between the two behaviors is not rigorously
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: Do we want to say something about slot based vs liquidity based here?

File: recommendations/local-resource-conservation.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+costs of running a node, funding channels on-chain and the opportunity cost of 
+capital committed to the attack.
+
+This document outlines recommendations for implementing local resource 
+conservation to mitigate slow jamming attacks. It is part of a hybrid solution 
+to mitigating channel jamming, intended to be followed by the introduction of 
+unconditional fees to mitigate fast jamming.
+
+## Overview
+Local resource conservation combines the follow elements to provide nodes with
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: following?

File: recommendations/local-resource-conservation.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+  * The corresponding outgoing HTLC (if present) will be forwarded with 
+    `endorsed` set to `1`.
+* Otherwise: 
+  * The HTLC will be limited to the remaining "general" slots and liquidity, 
+    and will be failed if there are no resources remaining in this bucket.
+  * The corresponding outgoing HTLC (if present) will be forwarded with 
+    `endorsed` set to `0`.
+
+In the steady state when Lightning is being used as expected, there should be 
+no need for use of the protected slots as payments are resolved quickly. Should
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: > as payments are resolved quickly

Maybe as the full capacity of slots and liquidity is rarely used.

File: recommendations/local-resource-conservation.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+  typically 2016 blocks (~2 weeks) at the time of writing. 
+* Maximum route length: the longest route that a valid [onion_packet](../04-onion-routing.md#packet-structure) 
+  can facilitate, ~25 hops at the time of writing.
+
+It is reasonable to expect an adversary to "about turn" - to behave perfectly 
+to build up reputation, then alter their behavior to abuse it. For this 
+reason, in-flight HTLCs have a temporary negative impact on reputation until 
+they are resolved.
+
+Upon receipt of `update_add_htlc`, the local considers the following to
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: the local node?

File: recommendations/local-resource-conservation.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+* In-flight HTLCs risk: any endorsed HTLC that the incoming link has 
+  in-flight on the requested outgoing link negatively affect reputation, 
+  based on the assumption that they will be held until just before their 
+  expiry height.
+
+On a per-HTLC basis, the local node will classify the sending node's 
+reputation for the offered HTLC as follows: 
+* if `incoming_link_revenue - in_flight_risk >= outgoing_link_revenue`: 
+  * the sending node is considered to have *sufficient* reputation for the 
+    forward.
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: Maybe use "endorsement" or "use of protected liquidity/slots"?

File: recommendations/local-resource-conservation.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+bootstrap reputation. 
+
+#### Outgoing Link Revenue
+Outgoing link revenue measures the damage to the local node (ie, the loss in 
+forwarding fees) that a slow jamming attack can incur. For an individual link, 
+this is equal to the total revenue forwarded in both directions over the 
+maximum allowed HTLC hold time.
+
+We define the following parameters: 
+* `outgoing_link_window`: the largest cltv delta from the current block height
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: Do we want this to be the largest cltv delta, or do we want more flexibility with setting the window size?

File: recommendations/local-resource-conservation.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+count torwards their unforegable contribution to building reputation, as they 
+push fees to the local node. While the incoming link may be valuable as a sink 
+(or predominantly outgoing) link, the HTLCs that the local node has forwarded 
+outwards do not represent fees paid by a potential attacker.
+
+#### In Flight HTLC Risk
+Whenever a HTLC is forwarded, we assume that it will resolve with the worst 
+case off-chain resolution time and dock reputation accordingly. This decrease 
+will be temporary in the case of fast resolution, and preemptively slash 
+reputation in the case where the HTLC is used as part of a slow jamming attack.
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md:234]: Do we do this before or after forwarding the HTLC?

File: recommendations/local-resource-conservation.md [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+For each `update_add_htlc` proposed by an incoming link:
+* If `endorsed` is non-zero AND the incoming link has sufficient local 
+  reputation for the HTLC (see [Recommendations for Reputation Scoring](#recommendations-for-reputation-scoring)):
+  * SHOULD forward the HTLC as usual.
+  * SHOULD set `endorsed` to 1 in the outgoing `update_add_htlc`.
+* Otherwise: 
+  * SHOULD reduce the remote peer's `max_accepted_htlcs` by 
+    `protected_slot_count` for the purposes of the proposed HTLC.
+  * SHOULD reduce the `max_htlc_value_in_flight` by 
+    `protected_liquidity_portion` * `funding_satoshis`.
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: Not sure I understand this...

File: recommendations/local-resource-conservation.md [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+average to minimize the amount of data that needs to be stored per-link. In 
+flight HLTCs can be accounted for separately to this calculation, as the node 
+will already have data for these HTLCs available.
+
+Track the following values for each rolling window: 
+* `last_update`: stores the timestamp of the last update to the decaying 
+    average, expressed in seconds.
+* `decaying_average`: stores the value of the decaying average.
+* `decay_rate`: a constant rate of decay based on the rolling window chosen, 
+  calculated as: `((1/2)^(1/window_length_seconds/2))`.
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: can be `((1/2)^(2/window_length_seconds))`

File: recommendations/local-resource-conservation.md [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+
+To update the `decaying_average` at time `t`:
+* `decaying_average` = `decaying_average` * `decay_rate` ^ `last_update_diff`.
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md:306]: Maybe define `last_update_diff`?

File: recommendations/local-resource-conservation.md [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+When updating the `decaying_average` to add the `effective_fee` of a newly 
+resolved HTLC at timestamp `t`: 
+* MUST update `decaying_average` as described above.
+* `decaying_average` = `decaying_average` + `effective_fee`
+
+### Multiple Channels
+If the local node has multiple channels open with the incoming and outgoing 
+links involved in a HTLC forward, they can be grouped as follows: 
+* Incoming link: consider `incoming_link_revenue` across all channels with the 
+  peer, as the sum of all forwarded HTLCs represents the total contribution.
```
Comment (ClaraShk) [inline recommendations/local-resource-conservation.md]: With more channels, that can create more damage in a slot based attack. Not sure it this is a huge problem.

File: recommendations/local-resource-conservation.md (hunk 3 above)
Comment (carlaKC) [reply recommendations/local-resource-conservation.md]: slots/resources 
+ do not saturate the channel

File: recommendations/local-resource-conservation.md (hunk 5 above)
Comment (carlaKC) [reply recommendations/local-resource-conservation.md]: the forward / the offered HTLC.

File: recommendations/local-resource-conservation.md [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+is observable within the protocol.
+
+If granted full access to a node's liquidity and slots, the maximum amount of 
+damage that can be inflicted on the targeted node is bounded by the following 
+protocol constraints:
+* Maximum HTLC hold: the largest cltv delta from the current block height that 
+  a node will allow a HTLC to set before failing it with [expiry_too_far](../04-onion-routing.md#failure-messages),
+  typically 2016 blocks (~2 weeks) at the time of writing. 
+* Maximum route length: the longest route that a valid [onion_packet](../04-onion-routing.md#packet-structure) 
+  can facilitate, ~25 hops at the time of writing.
```
Comment (carlaKC) [inline recommendations/local-resource-conservation.md]: Cut this, no longer used.

File: recommendations/local-resource-conservation.md (hunk 8 above)
Comment (carlaKC) [reply recommendations/local-resource-conservation.md]: funding_satoshis/max_htlc_value_in_flight

File: recommendations/local-resource-conservation.md (hunk 11 above)
Comment (carlaKC) [reply recommendations/local-resource-conservation.md]: MAY but must look at all in flight across all channels if you do that.

**Pull request closed without merging**